sqlmesh/core/schema_diff.py

```diff
@@ -14,83 +17,366 @@ class SchemaDeltaOp(Enum):
+                for c in self.columns
+            ]
+        )
+
+
+class ColumnPosition(PydanticModel):
+    is_first: bool
+    is_last: bool
+    after: t.Optional[str] = None
+
+    @classmethod
+    def create_first(cls) -> ColumnPosition:
+        return cls(is_first=True, is_last=False, after=None)
+
+    @classmethod
+    def create_last(cls, after: t.Optional[str] = None) -> ColumnPosition:
+        return cls(is_first=False, is_last=True, after=after)
+
+    @classmethod
+    def create_middle(cls, after: str) -> ColumnPosition:
+        return cls(is_first=False, is_last=False, after=after)
+
+    @classmethod
+    def create(
+        cls,
+        pos: int,
+        current_kwargs: t.List[exp.StructKwarg],
+        replacing_col: bool = False,
+    ) -> ColumnPosition:
+        is_first = pos == 0
+        is_last = pos == len(current_kwargs) - int(replacing_col)
+        after = None
+        if not is_first:
+            prior_kwarg = current_kwargs[pos - 1]
+            after, _ = _get_name_and_type(prior_kwarg)
+        return cls(is_first=is_first, is_last=is_last, after=after)
+
+    @property
+    def sqlglot_col_position(self) -> t.Optional[exp.ColumnPosition]:
+        column = exp.column(self.after) if self.after and not self.is_last else None
+        position = None
+        if self.is_first:
+            position = "FIRST"
+        elif column and not self.is_last:
+            position = "AFTER"
+        return exp.ColumnPosition(this=column, position=position)
+
 
 class SchemaDelta(PydanticModel):
     column_name: str
-    column_type: str
+    column_type: exp.DataType
     op: SchemaDeltaOp
+    parents: ParentColumns = ParentColumns.empty()
+    add_position: t.Optional[ColumnPosition] = None
+    current_type: t.Optional[exp.DataType] = None
 
     @classmethod
-    def add(cls, column_name: str, column_type: str) -> SchemaDelta:
-        return cls(column_name=column_name, column_type=column_type, op=SchemaDeltaOp.ADD)
+    def add(
+        cls,
+        column_name: str,
+        column_type: t.Union[str, exp.DataType],
+        position: ColumnPosition = ColumnPosition.create_last(),
+        parents: ParentColumns = ParentColumns.empty(),
+    ) -> SchemaDelta:
+        return cls(
+            column_name=column_name,
+            column_type=exp.DataType.build(column_type),
+            op=SchemaDeltaOp.ADD,
+            parents=parents,
+            add_position=position,
+        )
 
     @classmethod
-    def drop(cls, column_name: str, column_type: str) -> SchemaDelta:
-        return cls(column_name=column_name, column_type=column_type, op=SchemaDeltaOp.DROP)
+    def drop(
+        cls,
+        column_name: str,
+        column_type: t.Optional[t.Union[str, exp.DataType]] = None,
+        parents: ParentColumns = ParentColumns.empty(),
+    ) -> SchemaDelta:
+        column_type = exp.DataType.build(column_type) if column_type else exp.DataType.build("INT")
```

Comment:
no dialect here? you need to ensure that the inputs are sqlglot generic

Reply:
Yeah so I was thinking if someone wanted to make sure their types were parsed with a specific dialect then they would just pass in a DataType object where they used that dialect.